### PR TITLE
Windows: Support multi-channel output

### PR DESCRIPTION
**Pull request by jim-signal**
This PR is a follow-up to https://github.com/signalapp/webrtc/pull/68 in order to support multi-channel output devices on Windows.

The same approach is used, except in reverse. If the device's preferred format has more than 2 channels, an attempt will be made to see if the default 2 channel audio can be used. If not, a second attempt will be made with the device's preferred number of channels. When data is to be rendered, the audio frames will be converted from 2-channel stereo to N channels, with all channels after the stereo pair being silent.

When trying to initialize with N channels, the Channel Mask has to match what is expected for some devices. This isn't otherwise used in the ADM code though, other than keeping with the assumption that the first two channels are always LEFT and RIGHT. To support this, the mix_format is obtained in the Init() function so that it can be accessed. This would also affect multi-channel input, but testing of that passed successfully.

Some logs were also adjusted. The mix_format will always be logged and the secondary match operations will be logged a little more succinctly. Some other debug logs that were warnings or errors have been changed to always get logged.

Many multi-channel devices prefer the 32-bit floating point sample representation, but the 16-bit linear format that we want has been consistently accepted by WASAPI. According to [this](https://docs.microsoft.com/en-us/windows/win32/coreaudio/device-formats), the audio engine or device driver can do the conversion if necessary.

Finally, it should be noted that some multi-channel devices worked without the conversion. WebRTC was able to handle output devices having 8 channels in one case. However, it is not clear if this is truly supported or not given checks like this:

https://github.com/signalapp/webrtc/blob/e523aa4470ae6af8f1160996999aa4562eb1eb69/audio/audio_transport_impl.cc#L207

### CODE REVIEW

File: modules/audio_device/win/core_audio_base_win.cc [hunk 1]
```diff
@@ -388,15 +388,27 @@ bool CoreAudioBase::Init() {
...
+  WAVEFORMATPCMEX mix_format;
+  HRESULT res = core_audio_utility::GetSharedModeMixFormat(
+      audio_client.Get(), &mix_format);
+  if (FAILED(res)) {
+    return false;
+  }
+
+  RTC_LOG(LS_WARNING) << (IsInput() ? "input " : "output ")
+                      << "mix_format: "
+                      << core_audio_utility::WaveFormatToString(&mix_format);
```
Comment (jrose-signal) [inline modules/audio_device/win/core_audio_base_win.cc:401]: Nitpick: this should probably be `LS_INFO` since it's on the happy path.

File: modules/audio_device/win/core_audio_output_win.cc [hunk 2]
```diff
@@ -337,9 +341,29 @@ bool CoreAudioOutput::OnDataCallback(uint64_t device_frequency) {
...
+  if (format_.Format.nChannels > 2) {
+    UINT32 bytes_per_frame = format_.Format.nChannels * sizeof(int16_t);
+    UINT32 bytes_per_stereo_frame = 2 * sizeof(int16_t);
+
+    // Shift each stereo frame to start on the render frame's stride and
+    // set the remaining channels to zero (silence).
+    int16_t* audio_data_ptr = (int16_t*)audio_data;
+    for (UINT32 i = num_requested_frames - 1; i > 0; --i) {
+      memcpy(&audio_data_ptr[i * format_.Format.nChannels],
+             &audio_data_ptr[i * 2], bytes_per_stereo_frame);
```
Comment (jrose-signal) [inline modules/audio_device/win/core_audio_output_win.cc:360]: Nitpick: potentially-overlapping copies should be done with `memmove`, not `memcpy` (this is overlapping for i = 1|n = 3).

…but moreover you're doing a backwards loop rather than a forwards loop and that destroys data; this kind of loop has to be done forwards when shrinking in-place.

File: modules/audio_device/win/core_audio_output_win.cc [hunk 3]
```diff
@@ -337,9 +341,29 @@ bool CoreAudioOutput::OnDataCallback(uint64_t device_frequency) {
...
+      memset(&audio_data_ptr[i * format_.Format.nChannels + 2],
+             0, bytes_per_frame - bytes_per_stereo_frame);
```
Comment (jrose-signal) [inline modules/audio_device/win/core_audio_output_win.cc:362]: This is unnecessary; the data is out of bounds for a two-channel buffer (or, in a forwards-loop, it will be overwritten).

File: modules/audio_device/win/core_audio_output_win.cc [hunk 4]
```diff
@@ -337,9 +341,29 @@ bool CoreAudioOutput::OnDataCallback(uint64_t device_frequency) {
...
+    }
+    memset(&audio_data_ptr[2], 0, bytes_per_frame - bytes_per_stereo_frame);
```
Comment (jrose-signal) [inline modules/audio_device/win/core_audio_output_win.cc:364]: Not sure what this is doing.

**Comment by jrose-signal** [comment]
(I read the loop backwards, let me try again.)

**Review by jrose-signal — APPROVED** [review]

File: modules/audio_device/win/core_audio_base_win.cc (hunk 1 above)
Comment (jim-signal) [reply modules/audio_device/win/core_audio_base_win.cc:401]: I would like to see it in logs for a while, to get a reference for other audio issues on Windows, that is why I have it at warning level for now.

File: modules/audio_device/win/core_audio_output_win.cc (hunk 2 above)
Comment (jim-signal) [reply modules/audio_device/win/core_audio_output_win.cc:360]: This isn't shrinking though, it is expanding ? I'll look at it again though.

File: modules/audio_device/win/core_audio_output_win.cc (hunk 3 above)
Comment (jim-signal) [reply modules/audio_device/win/core_audio_output_win.cc:362]: We copied the l/r channels out of all the extra channels and somehow the first frame needs to be reset (channels beyond the first two), so this was why it was there.

File: modules/audio_device/win/core_audio_output_win.cc (hunk 4 above)
Comment (jim-signal) [reply modules/audio_device/win/core_audio_output_win.cc:364]: We copied the l/r channels out of all the extra channels and somehow the first frame needs to be reset (channels beyond the first two), so this was why it was there.

File: modules/audio_device/win/core_audio_output_win.cc (hunk 2 above)
Comment (jrose-signal) [reply modules/audio_device/win/core_audio_output_win.cc:360]: Yeah, I read the loop backwards, but "i = 1|n=3" is still overlapping.

**Comment by jim-signal** [comment]
> (I read the loop backwards, let me try again.)

Oh! Ok, thanks.

**Review by rashad-signal — APPROVED** [review]
